### PR TITLE
JMH Prototype 

### PR DESCRIPTION
**Pull request by metasim**
In response to #2348, this is an integration of the `sbt-jmh` plugin as a custom build plugin with some additional configuration keys for using and configuring benchmarking in whatever sub-project wishes to use it. Also created a `bench` sub-project to serve as an example of how to use it, and as a start for adding global benchmarking code (such as porting [legacy benchmarks](https://github.com/geotrellis/geotrellis-benchmark/tree/master/geotrellis/src/test/scala/geotrellis)).

Also committed the results from a run based on MergeQueue in an `archive` directory, providing an example of keeping historical results around.

### CODE REVIEW

**Comment by metasim** [comment]
Plot of MergeQueue vs. alternate implementation by @mteldridge in several multi-axis test configurations. See `geotrellis.spark.io.index.MergeQueueBench` in the `bench` subproject for details.

screen shot 2017-09-09 at 8 58 34 pm

**Comment by lossyrob** [comment]
This is looking great.

If I'm reading the graph correctly, that means @mteldridge's #2358 is a winner? Can we include a benchmark that tests the presort with the current MergeQueue as well?

**Comment by metasim** [comment]
@lossyrob The "shuffle" param indicates whether the data was randomized (true) or already sorted (false). "span" is how wide the interval is, and "skip" is the distance between the starting points of subsequent intervals. Varying these triggers overlapping vs. non-overlapping cases.

 That said, I welcome a review of my methodology. This is my first time really trying to measure this kind of thing.

JMH is pretty darn cool. The documentation sucks, but the examples are great and the features are impressive.

**Comment by metasim** [comment]
PS: I've ported a couple of the old benchmarks. So far it's pretty straightforward.

**Comment by fosskers** [comment]
I can review this today.

File: bench/build.sbt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+enablePlugins(GTBenchmarkPlugin)
+
+libraryDependencies += "org.slf4j" % "slf4j-nop" % "1.7.25"
+
+jmhIterations := Some(5)
+jmhTimeUnit := Some("ms")
+jmhExtraOptions := Some("-jvmArgsAppend -Xmx8G")
```
Comment (fosskers) [inline bench/build.sbt]: By hard coding these, we're assuming we'd never want to run the benches via say `jmh:run -t 1 -f 1 -wi 10 -i 10 .*Bench.*`? The `jmh:run` approach is definitely more "mystical".

File: bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import geotrellis.bench.init
+import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.infra.BenchmarkParams
+
+import scala.util.Random
+
+
+@BenchmarkMode(Array(Mode.AverageTime))
+@State(Scope.Thread)
+@Threads(4)
```
Comment (fosskers) [inline bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala]: Always 4?

File: bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      case "genericRasterMap" ⇒
+        genericRaster = new GRaster(init(len)(Random.nextInt))
+      case "rasterMap" ⇒
+        tile = ArrayTile(init(len)(Random.nextInt), size, size)
+      case _ ⇒ throw new MatchError("Have a new benchmark without initialization?")
+    }
+  }
+
+  @Benchmark
+  def genericRasterMap = {
```
Comment (fosskers) [inline bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala]: Return type here and elsewhere for the other functions.

File: bench/src/main/scala/geotrellis/raster/render/RenderBench.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      }).toMap
+    cmap =
+      ColorMap(map)
+
+    colorEncoding =
+      PngColorEncoding(cmap.colors, cmap.options.noDataColor, cmap.options.fallbackColor)
+  }
+
+  @Benchmark
+  def rendering = cmap.render(tile)
```
Comment (fosskers) [inline bench/src/main/scala/geotrellis/raster/render/RenderBench.scala]: Return type here too.

File: bench/build.sbt (hunk 1 above)
Comment (metasim) [reply bench/build.sbt]: You can still do `jmh:run -t 1 -f 1 -wi 10 -i 10 .*Bench.*` if you want to; the plugin just allows some of these things to be set and managed individually so you could do:

```
sbt> set jmhIterations := Some(5)
sbt> bench
```

File: bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala (hunk 2 above)
Comment (metasim) [reply bench/src/main/scala/geotrellis/raster/GenericRasterBench.scala]: Can't remember why I did that. I'll either remove or comment why the reason.

File: bench/build.sbt (hunk 1 above)
Comment (metasim) [reply bench/build.sbt]: PS: I think it's important to have a certain set of defaults version controlled to ensure you get somewhat consistent runs between analysis sessions (although the machine you run probably influences things heavily, so maybe it's false assumption of control).

**Comment by metasim** [comment]
@fosskers This is my first time using the review feature, so let me know if there's something I'm supposed to do when I'm done with the updates.

File: bench/build.sbt (hunk 1 above)
Comment (fosskers) [reply bench/build.sbt]: I think it's good to have sensible defaults.

**Comment by fosskers** [comment]
@metasim Nope, it's just a matter of me looking over the changes again and hitting "Approve". Looks great!

**Comment by fosskers** [comment]
How much of `geotrellis-benchmark` got ported here? Can we deprecate that repo yet?

**Comment by metasim** [comment]
I only ported a handful to work out the patterns involved. I think we should create a new ticket for porting the rest (or an identified subset) over.

**Pull request merged**